### PR TITLE
Example of the use of Substra with HAM10000 dataset

### PR DESCRIPTION
**Pull request by celinejacques**

### CODE REVIEW

File: HAM/data_scientist/assets/algo/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# this base image works in both CPU and GPU enabled environments
+FROM substrafoundation/substra-tools:0.7.0
+
+RUN mkdir /.cache
+RUN chmod -R 777 /.cache
```
Comment (natct10) [inline HAM/data_scientist/assets/algo/Dockerfile:5]: Simple question, I didn't test, but isn't 755 sufficient?

File: HAM/data_scientist/assets/algo/algo.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+from torch import optim, nn
+from torch.autograd import Variable
+
+logger = logging.getLogger(__name__)
+current_directory = os.path.dirname(__file__)
+
+# import sys
+# print(os.path.abspath('./'), flush=True)
+# sys.stdout = open('/sandbox/logs', 'w')
+# lzekhf
```
Comment (natct10) [inline HAM/data_scientist/assets/algo/algo.py]: Would you kindly remove this or clean it and uncomment it if it's useful?

File: HAM/data_owner/assets/dataset/opener.py [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+    #     for i in range(3):
+    #         pixels = imgs[:, :, i, :].ravel()
+    #         means.append(np.mean(pixels))
+    #         stds.append(np.std(pixels))
+
+    #     print("normMean = {}".format(means))
+    #     print("normStd = {}".format(stds))
+
+    #     return means, stds
```
Comment (natct10) [inline HAM/data_owner/assets/dataset/opener.py]: Why is this method commented?

File: HAM/data_scientist/assets/algo/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+
+RUN mkdir /model
+RUN chmod -R 777 /model
+
+# install dependencies
+RUN pip3 install pandas numpy sklearn pillow scipy 
+#tensorflow  pickle-mixin
```
Comment (natct10) [inline HAM/data_scientist/assets/algo/Dockerfile]: cleanup?

File: HAM/data_scientist/assets/algo/Dockerfile [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
+RUN pip3 install torch torchvision
+#RUN pip3 install keras==2.2.4
```
Comment (natct10) [inline HAM/data_scientist/assets/algo/Dockerfile]: It would be better to gather every dependencies in one requirements.txt file and to copy it in your Dockerfile

File: HAM/scripts/generate_data_samples.py [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+from glob import glob
+from sklearn.model_selection import KFold, train_test_split
+
+root_path = os.path.dirname(__file__)
+asset_de_path = os.path.join(root_path, '../HAM10000_DE/assets/')
+asset_ds_path = os.path.join(root_path, '../HAM10000_DS/assets/')
+
+# load dataset
+data_path = './data'
+# data_path = '/home/apricity/celine/HAM10000_skin_cancer/sources'
```
Comment (natct10) [inline HAM/scripts/generate_data_samples.py]: cleanup?

**Review by natct10 — CHANGES_REQUESTED** [review]
Thank yo for this great submission! I have added a few comments :+1:

File: HAM/data_owner/assets/dataset/opener.py (hunk 3 above)
Comment (celinejacques) [reply HAM/data_owner/assets/dataset/opener.py]: It was to normalize data but I don't use it in this example finally. I can remove it.

**Comment by celinejacques** [comment]
> Thank yo for this great submission! I have added a few comments 👍

Thanks for your comments! I clean up some files by removing some useless comments.

File: HAM/data_scientist/assets/algo/Dockerfile (hunk 1 above)
Comment (celinejacques) [reply HAM/data_scientist/assets/algo/Dockerfile:5]: I honestly don't know. However it would be worthwhile to test.

File: HAM/data_scientist/assets/algo/Dockerfile [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+FROM substrafoundation/substra-tools:0.7.0
+
+RUN mkdir /.cache
+RUN chmod -R 777 /.cache
+
+RUN mkdir /model
+RUN chmod -R 777 /model
+
+# install dependencies
+RUN pip3 install pandas numpy sklearn pillow scipy
```
Comment (Fabien-GELUS) [inline HAM/data_scientist/assets/algo/Dockerfile]: I don't think `sklearn pillow scipy `are required as they are not imported in the algo

File: HAM/data_scientist/assets/algo/algo.py [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+        model_ft.fc = nn.Linear(num_ftrs, num_classes)
+        input_size = 224
+        return model_ft, input_size
+
+
+    def train (self, X, y, models, rank):
+        
+        num_classes = 7
+        print(os.environ.get('AM_I_IN_A_DOCKER_CONTAINER', True))
+        if os.environ.get('AM_I_IN_A_DOCKER_CONTAINER', True):
```
Comment (Fabien-GELUS) [inline HAM/data_scientist/assets/algo/algo.py]: instead of checking if you're in a container, you could get the path of the algo folder with 
`root_path = os.path.dirname(__file__)` 
and with `ADD resnet.pth .` in your Dockerfile :
`weight_path = os.path.join(self.root_path,"resnet.pth")`

File: HAM/data_scientist/assets/dataset/opener.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
+"test"
```
Comment (Fabien-GELUS) [inline HAM/data_scientist/assets/dataset/opener.py:1]: this opener is the same as the data_owner one, except `shuffle=True`, is it necessary ? You could use one opener for both datasets

File: HAM/data_scientist/assets/objective/metrics.py [hunk 10]
```diff
@@ -0,0 +1,23 @@
...
+        """Returns the macro-average recall
+
+        :param y_true: actual values from test data
+        :type y_true: pd.DataFrame
+        :param y_true: predicted values from test data
+        :type y_pred: pd.DataFrame
+        :rtype: float
+        """
+
+        y_true = y_true[0].numpy()
```
Comment (Fabien-GELUS) [inline HAM/data_scientist/assets/objective/metrics.py]: why do you take only the first value ?

**Review by Fabien-GELUS — COMMENTED** [review]
This is a great example, and the first one with the debug mode ! Thank you for your contribution, I will test it soon !

File: HAM/data_scientist/assets/dataset/opener.py (hunk 9 above)
Comment (celinejacques) [reply HAM/data_scientist/assets/dataset/opener.py:1]: The idea with this example was also to simulate having two different datasets so I used two opener even if they are similar.

File: HAM/data_scientist/assets/algo/algo.py (hunk 8 above)
Comment (celinejacques) [reply HAM/data_scientist/assets/algo/algo.py]: Thanks! I'll try that and modify the code

File: HAM/data_owner/assets/dataset/Dockerfile [hunk 11]
```diff
@@ -0,0 +1 @@
+RUN chmod -R 777 /home/ubuntu/HAM/data/
```
Comment (natct10) [inline HAM/data_owner/assets/dataset/Dockerfile:1]:
```suggestion
RUN chmod -R 755 /home/ubuntu/HAM/data/

```

File: HAM/data_scientist/assets/algo/Dockerfile (hunk 1 above)
Comment (natct10) [reply HAM/data_scientist/assets/algo/Dockerfile:5]:
```suggestion
RUN chmod -R 755 /.cache
```

File: HAM/data_scientist/assets/algo/Dockerfile [hunk 12]
```diff
@@ -0,0 +1,19 @@
+# this base image works in both CPU and GPU enabled environments
+FROM substrafoundation/substra-tools:0.7.0
+
+RUN mkdir /.cache
+RUN chmod -R 777 /.cache
+
+RUN mkdir /model
+RUN chmod -R 777 /model
```
Comment (natct10) [inline HAM/data_scientist/assets/algo/Dockerfile:8]:
```suggestion
RUN chmod -R 755 /model
```

**Comment by RomainGoussault** [comment]
Closing this PR for inactivity.
A new repo with more examples will come up soon.
